keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1182 @@
+  discusses the problem of blocking PDBs and offers several workarounds. Similar to others also
+  offers a force deletion, but also a less destructive method of scaling up the application.
+  However, this also interferes with application deployment and has to be supported by the
+  application.
+
+To sum up. Some projects solve this by introducing validating admission webhooks. This has a couple
+of disadvantages. The webhooks are not easily discoverable by cluster admins. And they can block
+evictions for other applications if they are misconfigured or misbehave. It is not intended for the
+eviction API to be extensible in this way. The webhook approach is therefore not recommended.
+
+As seen in the experience reports and GitHub issues, some admins solve their problems by simply
+ignoring PDBs which can cause unnecessary disruptions or data loss. Some solve this by playing
+with the application deployment, but have to understand that the application supports this.
+
+### Goals
+- Kubectl drain should not evict and disrupt applications with max surge ability and instead
+  politely ask them to migrate their pods to another node or to remove them by creation of
+  NodeMaintenance object.
+- The PDB controller should detect and account for applications with max surge capability when
+  calculating PDB status.
+- Introduce a node maintenance controller that will help controllers like deployment controller
+  to migrate their pods.
+
+### Non-Goals
+- Introduce SurgeInterface, which CRDs could implement so that their pods would not be targeted
+  by node drain. Conceptually similar to ScaleInterface (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/scale/interfaces.go).
+- Introduce a field that could include non-critical daemon set pods
+  (priority `system-cluster-critical` or `system-node-critical`) for node maintenance/drain request.
+  The daemon set controller would then gracefully shut down these pods. Critical pods could be
+  overridden by the priority list mentioned below.
+- NodeMaintenance could include a plan of which pods to target first. Similar to graceful node
+  shutdown, we could include a list of priorities to decide which pods should be terminated first.
+  This list could optionally include pod timeouts, but could also wait for all the pods of a given
+  priority class to finish first without a timeout. This could also be used to target daemon set
+  pods of certain priorities (see point above). We could also introduce drain profiles based on
+  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
+  The logic for processing the decision list would be contained in the node maintenance controller,
+  which would set an intent to selected pods to shut down via the TargetedByNodeMaintenance
+  condition.
+- Introduce a node maintenance period or a TTL optional field as an upper bound on the duration of
```

Comment:
This might not be  a reliable way of delivering that information and there could be other better mechanism. But, if there is a sufficient interest, I think it can be added in later (for example suspended field in the spec).

Reply:
I have changed the NodeMaintenance API: now it should be possible to create the object without starting the actual maintenance. This can be used to announce the upcoming maintenance via `.spec.reason`.